Function units: mag, dB, and dex

mhvk opened the pull request:
Now that `Unit` and `Quantity` have calmed down a bit, something a bit bigger: introducing a framework for functional units and an application to logarithmic units, so that one do things like `dex(cm/s**2)`, `dB(mW)`, and `mag(AB)`. 

I think the usage I have in mind is best illustrated through an example -- see the output from an ipython-notebook session using magnitudes on http://www.astro.utoronto.ca/~mhvk/functional_units_magnitudes.html

For here, just a very brief outline:
```
In [1]: import astropy.units as u, astropy.units.functional as lu
In [2]: iv = lu.Magnitude(100. * u.ct/u.s)
In [3]: iv, iv.quantity, iv.unit
Out[3]: (<Magnitude -5.0 mag(ct / s)>, <Quantity 100.0 ct / s>, MagUnit('ct / s'))

In [4]: iv.unit.physical_unit, iv.unit.functional_unit
Out[4]: (Unit("ct / s"), Unit("mag"))
```

I hope initially to get feedback mostly on whether this covers what usage people can foresee.

Some notes:
-  This is meant as a more general alternative to the `Magnitude` class of #1577. In practice, the idea is not dissimilar; the main difference is that rather than store a zero-point magnitude, the unit corresponding to mag=0 is stored (or, equivalently, what one divides by before taking the `log`).
- Like #1577, this explicitly is about *units*, not magnitude systems (see discussion in #1577, which convinced me that this is the right approach).
- There are a lot of test cases, but no documentation yet, as I want to be sure people agree with the approach.
- Similarly, I have not yet worried about reading in functional units from strings, or writing out to formats other than `generic` (i.e., no FITS, etc.)
- I decided not to subclass from `UnitBase`, as that required much larger changes to `units` (see my earlier trial, https://github.com/mhvk/astropy/tree/units-functional). However, an abstract base class with expectations of what any unit should provide might be helpful. Technical advice very welcome!
- The commits before https://github.com/mhvk/astropy/commit/9bcf3b4f7d2a77e50b6d6fca2f9b2908ed192156 are from #1628, and only introduce `dex`, `mag`, and `dB` (without underlying physical unit).

mhvk commented:
As it doesn't look like functional units can make it in for 1.0, I removed the milestone.

mhvk commented:
@eteq, @mdboom - I rebased once more, and, as you'll have seen, sent off a request on astropy-dev for further review. I hope you will have time as well.  Any ideas about the larger issues mentioned above (https://github.com/astropy/astropy/pull/1894#issuecomment-68072477) would be welcome as well (though hopefully for a separate PR, this one is big enough as is!).

PBarmby commented:
Looks like this will be a great addition to `units`! 

I agree with @eteq above that  "functional unit" is confusing terminology, but it looks like it only appears in the code, and not in the documentation (unless directly referencing the code) -- is that right? Would it be less confusing to change from `astropy.units.function` to `astropy.units.log` ?

Comments on the documentation:
* generally very clear and well-done!
* do ST and AB mag systems need a definition reference? 
* in the example showing computation of reddening, would it be useful to explain why one of the 3 values at the end is `Magnitude`  and the other 2 are `Quantity`
* the example after "Note that one can take the automatic unit conversion quite far," while very nifty, may overwhelm a new user. 
* the "Numpy functions" section should more explicitly spell out which functions will work (statistical only?)

mhvk commented:
@PBarmby - thanks! I have now replaced any remaining instances of "functional unit" with "function unit". It may still be a little confusing to have `astropy.units.function` instead of `astropy.units.logarithmic`, but in practice one would anyway just do `from astropy.units import Magnitude`.

I should perhaps add that the reason I started like this was that `units/core.py` has a comment "TODO: Support functional units, e.g. log(x), ln(x)" -- so I assumed there were other functions that one might use. @eteq gave on example of something actually used in astronomy: the SDSS magnitudes which are defined using an `asinh` function (but which this PR does not yet support).

Thanks also for the comments on the documentation:
- Added references and also gave `STmag` and `ABmag` their own docstring.
- Now explicitly tell why the extinctions are quantities rather than magnitudes.
- Agreed that the final part may be too much for starting users, but I hope it is OK given that I introduce it as a separate note.
- Realised it is actually not that easy to explicitly list numpy functions and methods that are allowed/disallowed, and I worry that any list will quickly be out of data. Instead, I now added a note asking users to let us know if things work in unexpected ways.

mhvk commented:
In the spirit of the mailing list discussion, I removed myself as "assignee".  @mdboom - I fear this would be one for you... (though, @eteq, I'd definitely value further comments from you as well). Both: see https://github.com/astropy/astropy/pull/1894#issuecomment-68072477 for things that maybe should become separate issues.

mhvk commented on astropy/units/function/logarithmic.py:
```diff
@@ -0,0 +1,297 @@
+
+
+class LogUnit(FunctionUnitBase):
+    """Logarithmic unit containing a physical one
```
@mdboom's https://github.com/astropy/astropy/pull/3554#discussion_r26965955: add period.

mdboom commented on astropy/units/core.py:
```diff
@@ -94,8 +93,8 @@ def _normalize_equivalencies(equivalencies):
                 "Invalid equivalence entry {0}: {1!r}".format(i, equiv))
-        if not (isinstance(funit, UnitBase) and
-                (isinstance(tunit, UnitBase) or tunit is None) and
+        if not (funit is Unit(funit) and
```
I see.  I think what's here probably makes sense for now.

mhvk commented:
Nice to see it in!  It does mean i better get going with getting these units written to fits, etc. So, here again my list of leftovers, but now  with corresponding PRs:
1. The function units need to connected to the string readers so that `u.Unit('mag(Jy)')` would return a `FunctionUnit`.  --> see #3618.
2. Should support output to fits and vounit for the subset of all possible logarithmic units for which it is possible.  --> #3822
3. I had to do some duck-typing in `UnitBase` and `Unit` to ensure that function units get recognised when they should be, but not when the code expects a normal unit. A minimal abstract base class for units would be a more elegant solution. --> #3650

mdboom commented:
Sorry -- I skimmed through the PR and missed that there was still work left to do.  I can hit the big "revert" button if you'd rather not be rushed about the remaining things...

mhvk commented:
I actually think it is fine to have it in as is; this PR was rather big and therefore unwieldy. The most important follow-up is that which makes using the units easier by doing the parsing. While this is not essential, I think it is important enough to help users that I set the milestone for #3554 to 1.10. The other PRs are icing on the cake or restructuring, so they can be a bit slower.
